rasa/nlu/featurizers/count_vectors_featurizer.py

```diff
@@ -220,51 +245,111 @@ def train(
             # create spacy lemma_ for OOV_words
             self.OOV_words = [t.lemma_ for w in self.OOV_words for t in spacy_nlp(w)]
 
-        self.vectorizer = CountVectorizer(
-            token_pattern=self.token_pattern,
-            strip_accents=self.strip_accents,
-            lowercase=self.lowercase,
-            stop_words=self.stop_words,
-            ngram_range=(self.min_ngram, self.max_ngram),
-            max_df=self.max_df,
-            min_df=self.min_df,
-            max_features=self.max_features,
-            analyzer=self.analyzer,
-        )
-
-        lem_exs = [
-            self._get_message_text(example) for example in training_data.intent_examples
-        ]
-
-        self._check_OOV_present(lem_exs)
-
+        if self.use_shared_vocab:
+            shared_vectorizer = create_vectorizer(
+                self.token_pattern,
+                self.strip_accents,
+                self.lowercase,
+                self.stop_words,
+                (self.min_ngram, self.max_ngram),
+                self.max_df,
+                self.min_df,
+                self.max_features,
+                self.analyzer,
+            )
+            self.vectorizer = {
+                attribute: shared_vectorizer for attribute in MESSAGE_ATTRIBUTES
+            }
+        else:
+            self.vectorizer = {
+                attribute: create_vectorizer(
+                    self.token_pattern,
+                    self.strip_accents,
+                    self.lowercase,
+                    self.stop_words,
+                    (self.min_ngram, self.max_ngram),
+                    self.max_df,
+                    self.min_df,
+                    self.max_features,
+                    self.analyzer,
+                )
+                for attribute in MESSAGE_ATTRIBUTES
+            }
+
+        cleaned_attribute_texts = {}
+
+        for attribute in MESSAGE_ATTRIBUTES:
+            lem = [
+                self._get_message_text_by_attribute(example, attribute)
+                for example in training_data.intent_examples
+            ]
+            self._check_OOV_present(lem)
+            cleaned_attribute_texts[attribute] = lem
+
+        combined_cleaned_texts = []
+        if self.use_shared_vocab:
+            for attribute in MESSAGE_ATTRIBUTES:
+                combined_cleaned_texts += cleaned_attribute_texts[attribute]
+
+        featurized_attributes = {}
+        # noinspection PyPep8Naming
         try:
-            # noinspection PyPep8Naming
-            X = self.vectorizer.fit_transform(lem_exs).toarray()
+            for attribute in MESSAGE_ATTRIBUTES:
+                if self.use_shared_vocab:
+                    self.vectorizer[attribute].fit(combined_cleaned_texts)
```

Comment:
Now `attribute` is something else. Do I miss something?

Reply:
also this operation is done three times unnecessary